index.d.ts

```diff
@@ -1,135 +1,96 @@
-declare module "before-after-hook" {
-  export interface HookCollection {
-    /**
-     * Invoke before and after hooks.
-     */
-    (name: string | string[], method: (options: any) => Promise<any> | any): Promise<any>
-    /**
-     * Invoke before and after hooks.
-     */
-    (name: string | string[], method: (options: any) => Promise<any> | any, options: any): Promise<any>
-    /**
-     * Add before hook for given name. Returns `hook` instance for chaining.
-     */
-    before (name: string, method: (options: any) => Promise<any> | any): HookCollection
-    /**
-     * Add error hook for given name. Returns `hook` instance for chaining.
-     */
-    error (name: string, method: (options: any) => Promise<any> | any): HookCollection
-    /**
-     * Add after hook for given name. Returns `hook` instance for chaining.
-     */
-    after (name: string, method: (options: any) => Promise<any> | any): HookCollection
-    /**
-     * Add wrap hook for given name. Returns `hook` instance for chaining.
-     */
-    wrap (name: string, method: (options: any) => Promise<any> | any): HookCollection
-    /**
-     * Removes hook for given name. Returns `hook` instance for chaining.
-     */
-    remove (name: string, beforeHookMethod: (options: any) => Promise<any> | any): HookCollection
-  }
+type HookMethod<O, R> = (options: O) => R | Promise<R>
 
-  export interface HookSingular<T> {
-    /**
-     * Invoke before and after hooks without options
-     */
-    (method: (options: T) => T | null | void): Promise<T>
-    /**
-     * Invoke before and after hooks without options
-     */
-    (method: (options: T) => Promise<T | null | void>): Promise<T>
-    /**
-     * Invoke before and after hooks with options
-     */
-    (method: (options: T) => T | null | void, options: T): Promise<T>
-    /**
-     * Invoke before and after hooks with options
-     */
-    (method: (options: T) => Promise<T | null | void>, options: T): Promise<T>
+type BeforeHook<O> = (options: O) => void
+type ErrorHook<O, E> = (error: E, options: O) => void
```

Comment:
For the unexported types (i.e. `HookMethod`, `BeforeHook`, `ErrorHook`, `AfterHook`, `WrapHook`, `AnyHook`) it doesn't matter. The user will never interact with them. They are just implementation details.

For the exported types (i.e. `HookSingular`and `Singular`), they have the generic signature of `<O, R, E>`. The idea was to arrange by usage frequency.
- Most of the time only the Options is typed. So the user can just provide that and omit the remaining two on `new Signular<O>()` call and the rest will fallback to `any`.
- Also, users who are already using the previous typings, this new one won't be breaking those codes. Because `O` (`T` in the previous typings) was the first generic.

If you want, I can rearrange them. Which order would you prefer?
- `<R, E, O>`
- `<E, R, O>`

But
- This will break the already existing codes by the users.
- `new Singular<>()` typing has to provide both `R` & `E` in order to strictly type the Options. So, instead of `new Signular<O>()`, it will be `new Signular<any, any, O>()`.

Reply:
That all makes perfectly sense. Also this is Typescript and folks using that will get context information anyway most of the time. Thanks for laying it out for me, much appreciated 👍